src/psyclone/psyGen.py

```diff
@@ -2784,6 +2796,108 @@ def validate(self, node, options=None):
+        the option'''
+        valid_options = type(self).get_valid_options()
+        if option_name not in valid_options.keys():
+            raise ValueError(f"'{type(self).__name__}' failed to get option "
+                             f"'{option_name}' as it is not provided as a "
+                             f"keyword argument to the apply method.")
+        return kwargs.get(option_name, valid_options[option_name]['default'])
+
+    @classmethod
+    def get_valid_options(cls):
+        '''
+        Pulls the valid options from the apply method. It also recurses
+        upwards to the superclasses of this transformation and pulls
+        their valid options as well if they exist.
+        '''
+        valid_options = OrderedDict()
+        # Loop through the inherited classes of this class, starting with the
+        # highest superclass. For some reason super isn't working here.
+        for base_cls in cls.__mro__[::-1]:
+            base_cls_apply = base_cls.__dict__.get("apply", None)
+            if base_cls_apply is None:
+                continue
+            signature = inspect.signature(base_cls_apply)
+            # Loop over the arguments to the apply call.
+            for k, v in signature.parameters.items():
+                if k == "options":
+                    continue
+                # If the argument is a keyword argument, i.e. it has a default
+                # value then we add it to the list of options.
+                if v.default is not inspect.Parameter.empty:
+                    valid_options[k] = {}
+                    valid_options[k]['default'] = v.default
+                else:
+                    # If its not a keyword argument then we skip it.
+                    continue
+                # If there is type information on the keyword argument, then we
+                # store the information so we can generate docs and do
+                # automatic type checking of options.
+                if v.annotation is not inspect.Parameter.empty:
+                    valid_options[k]['type'] = v.annotation
+                    valid_options[k]['typename'] = stringify_annotation(
+                            v.annotation
+                    )
+                else:
+                    valid_options[k]['type'] = None
+                    valid_options[k]['typename'] = None
+        return valid_options
+
+    def validate_options(self, **kwargs):
+        '''
+        Checks the options provided to this transformation are valid. Makes
+        use of get_valid_options to work out the available options.
+
+        :raises ValueError: If an invalid option was provided.
+        :raises TypeError: If an option was provided with a value of the wrong
+                           type.
+        '''
+        valid_options = type(self).get_valid_options()
+        invalid_options = []
+        wrong_types = {}
+        for option in kwargs:
+            if option == "options":
+                continue
```

Comment:
Presumably we'll also look into the `options` dict eventually?

Reply:
Not as planned for now, this is replacement for the `options` dict so at some point in the future i'd deprecate and remove the options dict in favour of the new strategy (thus removing the need for this check).

Can reevaluate if you want to keep both strategies, but I think its unnecessary when for most current options you can just pass `**my_dict` instead of `options=mydict` with no other changes.